src/modules/simulator/simulator_mavlink.cpp

```diff
@@ -371,22 +370,29 @@ void Simulator::handle_message_hil_sensor(const mavlink_message_t *msg)
 
 	update_sensors(now_us, imu);
 
+	static float battery_percentage = 1.0f;
+	static uint64_t last_integration_us = 0;
+
 	// battery simulation (limit update to 100Hz)
 	if (hrt_elapsed_time(&_battery_status.timestamp) >= 10_ms) {
 
 		const float discharge_interval_us = _param_sim_bat_drain.get() * 1000 * 1000;
 
 		bool armed = (_vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_ARMED);
 
-		if (!armed || batt_sim_start == 0 || batt_sim_start > now_us) {
-			batt_sim_start = now_us;
+		if (armed && last_integration_us != 0) {
+			battery_percentage -= (now_us - last_integration_us) / discharge_interval_us;
+			last_integration_us = now_us;
+
+		} else if (armed) {
+			last_integration_us = now_us;
+
+		} else {
+			last_integration_us = 0;
```

Comment:
Hi @julianoes , would you be interested in doing this here instead?
```
if (armed) {
	if (last_integration_us != 0) {
		battery_percentage -= (now_us - last_integration_us) / discharge_interval_us;
	}

	last_integration_us = now_us;

} else {
	last_integration_us = 0;
}
```

Everything else looks good to me!